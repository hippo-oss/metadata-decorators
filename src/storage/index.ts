export * from './storage.api';
export * from './storage.constants';
