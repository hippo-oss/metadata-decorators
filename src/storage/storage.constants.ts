export const METADATA_KEY = '@hippo-oss/dto-decorators';
