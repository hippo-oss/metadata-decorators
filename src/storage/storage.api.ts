import { Target } from '@hippo-oss/dto-decorators';
import 'reflect-metadata';

import { DTOSchema, DTOSchemaProperty } from '../types';
import { METADATA_KEY } from './storage.constants';

/* Save schema metadata.
 */
export function putSchema<T>(
    target: Target,
    schema: DTOSchema<T>,
): void {
    Reflect.defineMetadata(
        METADATA_KEY,
        schema,
        target,
    );
}

/* Get schema metadata.
 */
export function getSchema<T>(
    target: Target,
): DTOSchema<T> | undefined {
    const schema = Reflect.getMetadata(
        METADATA_KEY,
        target,
    ) as unknown;

    if (schema === undefined) {
        return undefined;
    }

    return schema as DTOSchema<T>;
}

/* Save option metadata.
 */
export function putProperty<T>(
    target: Target,
    propertyKey: string | symbol,
    property: DTOSchemaProperty,
): void {
    const key = String(propertyKey) as keyof T;

    const schema = getSchema<T>(target) ?? {} as DTOSchema<T>;
    schema[key] = property;

    putSchema(target, schema);
}

/* Retrieve saved option metadata.
 */
export function getProperty<T>(
    target: Target,
    propertyKey: string | symbol,
): DTOSchemaProperty | undefined {
    const key = String(propertyKey) as keyof T;

    const schema = getSchema<T>(target);

    if (schema === undefined || schema[key] === undefined) {
        return undefined;
    }

    return schema[key] as DTOSchemaProperty;
}
