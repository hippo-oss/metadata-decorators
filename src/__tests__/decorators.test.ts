import { getSchema } from '../storage';
import { Address, Location } from './fixtures';

describe('decorators', () => {
    it('define Address properties', () => {
        const properties = getSchema(Address);
        expect(properties).toMatchSnapshot();
    });
    it('define Location properties', () => {
        const properties = getSchema(Location);
        expect(properties).toMatchSnapshot();
    });
});
