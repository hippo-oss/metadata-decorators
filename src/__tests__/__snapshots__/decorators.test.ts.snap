// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators define Address properties 1`] = `
Object {
  "city": Object {
    "name": "IsString",
    "options": Object {
      "description": "The address's city.",
      "example": "Washington",
      "optional": false,
    },
  },
  "id": Object {
    "name": "IsUUID",
    "options": Object {
      "description": "The address's persistent id",
      "example": "00000000-0000-4000-8000-000000000000",
      "optional": true,
    },
  },
  "location": Object {
    "name": "IsNested",
    "options": Object {
      "description": "The location of the address.",
      "optional": true,
      "type": [Function],
    },
  },
  "state": Object {
    "name": "IsEnum",
    "options": Object {
      "description": "The address's state.",
      "enum": Object {
        "CA": "CA",
        "DC": "DC",
        "TX": "TX",
      },
      "enumName": "State",
      "example": "DC",
      "optional": false,
    },
  },
  "street": Object {
    "name": "IsString",
    "options": Object {
      "description": "The address's street location",
      "example": "1600 Pennsylvania Ave",
      "optional": false,
    },
  },
  "zipcode": Object {
    "name": "IsString",
    "options": Object {
      "description": "The address's zipcode.",
      "example": "20500-0004",
      "optional": false,
      "pattern": /\\[0-9\\]\\{5\\}\\(-\\[0-9\\]\\{4\\}\\)\\?/,
    },
  },
}
`;

exports[`decorators define Location properties 1`] = `
Object {
  "latitude": Object {
    "name": "IsNumber",
    "options": Object {
      "description": "The latitude of the location",
      "example": 38.89789,
      "maxValue": 90,
      "minValue": -90,
      "optional": false,
    },
  },
  "longitude": Object {
    "name": "IsNumber",
    "options": Object {
      "description": "The longitude of location.",
      "example": -77.03651,
      "maxValue": 180,
      "minValue": -180,
      "optional": true,
    },
  },
}
`;
